finality-aleph/src/network/manager/service.rs

```diff
@@ -42,13 +42,34 @@ struct Session<D: Data> {
 }
 
 #[derive(Clone)]
-struct PreSession {
+struct PreValidatorSession {
     session_id: SessionId,
     verifier: AuthorityVerifier,
     node_id: NodeIndex,
     pen: AuthorityPen,
 }
 
+#[derive(Clone)]
+struct PreNonvalidatorSession {
```

Comment:
Not sure what's good practice in Rust in this case, but in other programming languages, I worked in the past which has type inheritance, I'd consider making some base class with fields `session_id` and `verifier` as we can see both `PreValidatorSession` and `PreNonValidatorSession` have those in common - nevertheless inheritance as such is not a popular concept nowadays. In Rust, I understand we can use composition (like make base type `Validator` which have those two fields and make `Validator` a field of  `PreValidatorSession`), but it might be too much hustle for this little duplication.

Reply:
I think the duplication here is more coincidental than meaningful – even if we created a common substruct we couldn't use the commonality for anything in the rest of the code, because we do very different things with these two cases.

Or at least it would require a major refactor to actually use the commonality – if you have an idea for such broader restructuring I would be quite interested. It might be very viable after this round of changes, since now I'm no longer confident the design avoids major redundancies.